src/spawn.c

```diff
@@ -562,26 +566,106 @@ static gboolean spawn_async_with_pipes(const gchar *working_directory, const gch
 	g_message("full spawn command line: %s", command->str);
 #endif
 
-	while (envp && *envp)
+	while (envp && *envp && success)
 	{
-		g_array_append_vals(environment, *envp, strlen(*envp) + 1);
+		glong w_entry_len;
+		wchar_t *w_entry;
+		gchar *tmp = NULL;
+
+		// FIXME: remove this and rely on UTF-8 input
+		if (! g_utf8_validate(*envp, -1, NULL))
+		{
+			tmp = utils_get_utf8_from_locale(*envp);
```

Comment:
remembering that this does nothing on windows, is it really what is needed here?